File: 1LabGo/queue.go
package main

import "fmt"

type QueueNode struct {
	data string
	next *QueueNode
}

type Queue struct {
	head *QueueNode
	tail *QueueNode
	size int
}

func (q *Queue) QPush(data string) {
	newNode := &QueueNode{data: data}
	if q.head == nil {
		q.head = newNode
		q.tail = newNode
	} else {
		q.tail.next = newNode
		q.tail = newNode
	}
	q.size++
}

func (q *Queue) QPop() (string, error) {
	if q.head == nil {
		return "", fmt.Errorf("ошибка: очередь пуста")
	}
	data := q.head.data
	q.head = q.head.next
	if q.head == nil {
		q.tail = nil
	}
	q.size--
	return data, nil
}

func (q *Queue) QPeek() (string, error) {
	if q.head == nil {
		return "", fmt.Errorf("ошибка: очередь пуста")
	}
	return q.head.data, nil
}

func (q *Queue) QPrint() {
	if q.head == nil {
		fmt.Println("Очередь пуста.")
		return
	}
	fmt.Print("Начало -> ")
	current := q.head
	for current != nil {
		fmt.Printf(`"%s" `, current.data)
		current = current.next
	}
	fmt.Println("-> Конец")
}

File: 1Lab/Queue.cpp
#include "Queue.h"
#include <iostream>

using namespace std;

void QCREATE(Queue& q) {
    q.head = nullptr;
    q.tail = nullptr;
    q.size = 0;
    cout << "Очередь создана." << endl;
}

bool isEmpty(const Queue& q) {
    return q.head == nullptr;
}

void QPUSH(Queue& q, const string& value) {
    QNode* newNode = new QNode{value, nullptr};
    if (isEmpty(q)) {
        q.head = newNode;
        q.tail = newNode;
    } else {
        q.tail -> next = newNode;
        q.tail = newNode;
    }
    q.size++;
}

string QPOP(Queue& q) {
    if (isEmpty(q)) {
        cerr << "Ошибка: очередь пуста! Невозможно выполнить pop." << endl;
        return "";
    }
    QNode* temp = q.head;
    string popped_value = temp -> data;
    q.head = q.head -> next;

    if (q.head == nullptr) {
        q.tail = nullptr;
    }
    delete temp;
    q.size--;
    return popped_value;
}

string QPEEK(const Queue& q) {
    if (isEmpty(q)) {
        cerr << "Ошибка: очередь пуста! Невозможно прочитать элемент." << endl;
        return "";
    }
    return q.head->data;
}

void QPRINT(const Queue& q) {
    if (isEmpty(q)) {
        cout << "Очередь пуста." << endl;
        return;
    }

    cout << "Начало -> ";
    QNode* current = q.head;
    while (current != nullptr) {
        cout << "\"" << current->data << "\" ";
        current = current->next;
    }
    cout << "-> Конец" << endl;
}

void QCLEAR(Queue& q) {
    while (!isEmpty(q)) {
        QPOP(q);
    }
    cout << "Очередь очищена." << endl;
}